engines/wage/combat.cpp

```diff
@@ -0,0 +1,915 @@
+		Obj *obj = chr->_armor[Chr::SHIELD_ARMOR];
+
+		snprintf(buf, 512, "%s carries %s%s.", getGenderSpecificPronoun(chr->_gender, true),
+				obj->_namePlural ? "" : getIndefiniteArticle(obj->_name), obj->_name.c_str());
+		appendText(buf);
+	}
+}
+
+void WageEngine::performCombatAction(Chr *npc, Chr *player) {
+	if (npc->_context._frozen)
+		return;
+
+	RandomHat hat(_rnd);
+
+	bool winning = (npc->_context._statVariables[PHYS_HIT_CUR] > player->_context._statVariables[PHYS_HIT_CUR]);
+	int validMoves = getValidMoveDirections(npc);
+	ObjArray *weapons = npc->getWeapons(false);
+	ObjArray *magics = npc->getMagicalObjects();
+	// TODO: Figure out under what circumstances we need to add +1
+	// for the chance (e.g. only when all values were set to 0?).
+	if (winning) {
+		if (!_world->_weaponMenuDisabled) {
+			if (weapons->size() > 0)
+				hat.addTokens(kTokWeapons, npc->_winningWeapons + 1);
+			if (magics->size() > 0)
+				hat.addTokens(kTokMagic, npc->_winningMagic);
+		}
+		if (validMoves != 0)
+			hat.addTokens(kTokRun, npc->_winningRun + 1);
+		if (npc->_inventory.size())
+			hat.addTokens(kTokOffer, npc->_winningOffer + 1);
+	} else {
+		if (!_world->_weaponMenuDisabled) {
+			if (weapons->size() > 0)
+				hat.addTokens(kTokWeapons, npc->_losingWeapons + 1);
+			if (magics->size() > 0)
+				hat.addTokens(kTokMagic, npc->_losingMagic);
+		}
+		if (validMoves != 0)
+			hat.addTokens(kTokRun, npc->_losingRun + 1);
+		if (npc->_inventory.size())
+			hat.addTokens(kTokOffer, npc->_losingOffer + 1);
+	}
+
+	ObjList *objs = &npc->_currentScene->_objs;
+	if (npc->_inventory.size() < npc->_maximumCarriedObjects) {
+		int cnt = 0;
+		for (ObjList::const_iterator it = objs->begin(); it != objs->end(); ++it, ++cnt) {
+			if ((*it)->_type != Obj::IMMOBILE_OBJECT) {
+				// TODO: I'm not sure what the chance should be here.
+				hat.addTokens(cnt, 123);
+			}
+		}
+	}
+
+	int token = hat.drawToken();
+	switch (token) {
+		case kTokWeapons:
+			// TODO: I think the monster should choose the "best" weapon.
+			performAttack(npc, player, weapons->operator[](_rnd->getRandomNumber(weapons->size() - 1)));
+			break;
+		case kTokMagic:
+			// TODO: I think the monster should choose the "best" magic.
+			performMagic(npc, player, magics->operator[](_rnd->getRandomNumber(magics->size() - 1)));
+			break;
+		case kTokRun:
+			performMove(npc, validMoves);
+			break;
+		case kTokOffer:
+			performOffer(npc, player);
+			break;
+		case kTokNone:
+			break;
+		default:
+			{
+				int cnt = 0;
+				for (ObjList::const_iterator it = objs->begin(); it != objs->end(); ++it, ++cnt)
+					if (cnt == token)
+						performTake(npc, *it);
+				break;
+			}
+	}
+
+	delete weapons;
+	delete magics;
+}
+
+const char *targets[] = { "head", "chest", "side" };
+
+void WageEngine::performAttack(Chr *attacker, Chr *victim, Obj *weapon) {
+	if (_world->_weaponMenuDisabled)
+		return;
+
+	// TODO: verify that a player not aiming will always target the chest??
+	int targetIndex = -1;
```

Comment:
I assume that this comment refers to the case where targetIndex is -1, i.e. when weapon->_type is Obj::MAGICAL_OBJECT, and chance < attacker->_physicalAccuracy. Should there be an error message in this case? Or will out of memory access occur?